tools/find_panics.py

```diff
@@ -0,0 +1,453 @@
+abstract_origin_re = re.compile(r""".*DW_AT_abstract_origin.*""")
+dw_at_linkage_name_re = re.compile(r""".*DW_AT_linkage_name.*""")
+dw_at_name_re = re.compile(r""".*DW_AT_name.*""")
+
+
+def matches_panic_funcs(name):
+    """If the passed name contains one of the known panic_functions,
+    return the match
+    """
+    for func in panic_functions:
+        if func in name:
+            return func
+    return ""
+
+
+def linkage_or_origin_all_parents(elf, addr, linkage=False):
+    """Returns a list of the abstract origin or linkage of all parents of the dwarf
+    location for the passed address
+    """
+    result = subprocess.run(
+        (DWARFDUMP, "--lookup=0x" + addr, "-p", elf), capture_output=True, text=True
+    )
+    dwarfdump = result.stdout
+    regex = abstract_origin_re
+    if linkage:
+        regex = dw_at_linkage_name_re
+    matches = re.findall(regex, dwarfdump)
+
+    def getFunction(line):
+        return line.strip().split('"')[1]
+
+    origins = list(map(getFunction, matches))
+    return origins
+
+
+def any_origin_matches_panic_func(elf, addr):
+    """returns name if any origin for the passed addr matches one
+    of the functions in the panic_functions array
+    """
+    origins = linkage_or_origin_all_parents(elf, addr)
+    for origin in origins:
+        name = matches_panic_funcs(origin)
+        if name:
+            return name
+    return ""
+
+
+def any_linkage_matches_panic_func(elf, addr):
+    """returns True + name if any linkage for the passed addr matches one
+    of the functions in the panic_functions array
+    """
+    linkages = linkage_or_origin_all_parents(elf, addr, True)
+    for linkage in linkages:
+        name = matches_panic_funcs(linkage)
+        if name:
+            return name
+    return ""
+
+
+def check_for_source_in_parent(elf, addr):
+    """Takes in a dwarfdump lookup including parents of the source DWARF
+    location, returns the first parent with a call file not in
+    the core library. If found, this often indicates the source of the panic
+    in the Tock source code.
+    """
+    result = subprocess.run(
+        (DWARFDUMP, "--lookup=0x" + addr, "-p", elf), capture_output=True, text=True
+    )
+    dwarfdump = result.stdout
+    matches = re.findall(dw_at_file_re, dwarfdump)
+
+    def getFile(line):
+        return line.strip().split('"')[1]
+
+    source_files = list(map(getFile, matches))
+    for (i, f) in enumerate(source_files[::-1]):
+        if "/core/" not in f:
+            line_matches = re.findall(dw_at_line_re, dwarfdump)
+
+            def getLine(line):
+                return line.strip().split("(")[1].split(")")[0]
+
+            source_lines = list(map(getLine, line_matches))
+            source_line = source_lines[::-1][i]
+            return (f, source_line)
+    return ("", "")
+
+
+def parse_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("elf", help="ELF file for analysis")
```

Comment:
Usually variable arguments (rather than specific options) are all caps, so ELF. E.g.:

`Usage: tar [OPTION...] [FILE]...`

So think this should print out as

`usage: find_panics.py [-h] [--verbose] [--riscv] ELF`

Reply:
Thanks, fixed!